File: netcore/core/component.cpp
/****************************************************************************

Git <https://github.com/sniper00/MoonNetLua>
E-Mail <hanyongtao@live.com>
Copyright (c) 2015-2016 moon
Licensed under the MIT License <http://opensource.org/licenses/MIT>.

****************************************************************************/
#include "component.h"

namespace moon
{
	struct  component::component_imp
	{
		component_imp() noexcept
			:owner_(nullptr)
			,enable_update_(false)
		{

		}

		std::string	 name_;
		component* owner_;
		bool enable_update_;
		std::unordered_map<std::string, component_ptr_t>	components_;
	};

	component::component() noexcept
		:component_imp_(new component_imp)
	{
	}

	component::~component()
	{
		SAFE_DELETE(component_imp_);
	}
	
	const std::string & component::name()
	{
		return component_imp_->name_;
	}

	component* component::owner_imp()
	{
		return component_imp_->owner_;
	}

	void component::set_enable_update(bool v)
	{
		component_imp_->enable_update_ = v;
	}

	bool component::enable_update()
	{
		return component_imp_->enable_update_;
	}

	component_ptr_t component::get_component_imp(const std::string & name)
	{
		auto iter = component_imp_->components_.find(name);
		if (iter != component_imp_->components_.end())
		{
			return iter->second;
		}
		return nullptr;
	}

	bool component::add_component_imp(component_ptr_t v)
	{
		auto iter = component_imp_->components_.emplace(v->name(), v);
		assert(iter.second&&"The component is already exist!");
		return iter.second;
	}

	void component::traverse(std::function<void(component*)> cb)
	{
		for (auto& it : component_imp_->components_)
		{
			cb(it.second.get());
		}
	}

	void component::set_name(const std::string & name)
	{
		component_imp_->name_ = name;
	}

	void component::set_owner(component * v)
	{
		component_imp_->owner_ = v;
	}

	void component::start()
	{
		for (auto& it : component_imp_->components_)
		{
			it.second->start();
		}
	}

	void component::destory()
	{
		for (auto& it : component_imp_->components_)
		{
			it.second->destory();
		}
	}

	void component::update()
	{
		for (auto& it : component_imp_->components_)
		{
			if (it.second->enable_update())
			{
				it.second->update();
			}
		}
	}

}

File: resources/Common/Component.lua
require("functions")
require("Log")

local Component = class("Component")

function Component:ctor()
    self.components = { }
    self.name = "Unkonwn Component"
    self.owner = nil
    self.enableupdate = false

    Log.Trace("ctor Component")

end

function Component:GetName()
    return self.name
end

function Component:SetName(name)
    self.name = name
end

function Component:GetOwner()
    return self.owner
end

function Component:SetOwner(owner)
    self.owner = owner
end

function Component:AddComponent(name, cpt)
    assert(self.components[name] == nil, string.format("Component named [%s] already exist!", name))
    self.components[name] = cpt
end

function Component:GetComponent(name)
    assert(nil ~= name, "GetComponent name must not be nil")
    return self.components[name]
end

function Component:Start()
    for _, v in pairs(self.components) do
        v:Start()
    end
end

function Component:Update()
    for _, v in pairs(self.components) do
        if v:IsEnableUpdate() then
            v:Update()
        end
    end
end

function Component:Destory()
    for _, v in pairs(self.components) do
        v:Destory()
    end
end

function Component:SetEnableUpdate(b)
    self.enableupdate = b
end

function Component:IsEnableUpdate()
    return self.enableupdate
end

function Component:OnClientClose(player)
    for _, v in pairs(self.components) do
        v:OnClientClose(player)
    end
end

return Component